ArcGISRuntimeSDKQt_CppSamples/Layers/ListKmlContents/ListKmlContents.cpp

```diff
@@ -0,0 +1,313 @@
+
+  const QString filepath = defaultDataPath() + "/ArcGIS/Runtime/Data/kml/esri_test_data.kmz";
+
+  // create new KML layer
+  m_kmlDataset = new KmlDataset(filepath, this);
+  KmlLayer* kmlLayer = new KmlLayer(m_kmlDataset, this);
+  m_scene->operationalLayers()->append(kmlLayer);
+
+  connect(m_kmlDataset, &KmlDataset::doneLoading, this, [this](Error loadError)
+  {
+    if (!loadError.isEmpty())
+    {
+      qDebug() << loadError.message() << loadError.additionalMessage();
+      return;
+    }
+
+    // recursively build tree to display KML contents
+    for (KmlNode* node : m_kmlDataset->rootNodes())
+    {
+      m_kmlNodesList << node;
+      buildTree(node);
+    }
+
+    // if at top node, then display children
+    if (!m_kmlNodesList.isEmpty() && m_kmlNodesList[0]->parentNode() == nullptr)
+    {
+      displayChildren(m_kmlNodesList[0]);
+    }
+  });
+}
+
+QString ListKmlContents::getKmlNodeType(KmlNode *node)
+{
+  if (node == nullptr)
+    return "";
+
+  QString type = "";
+  switch (node->kmlNodeType()) {
+  case KmlNodeType::KmlDocument:
+    type = "KmlDocument";
+    break;
+  case KmlNodeType::KmlFolder:
+    type = "KmlFolder";
+    break;
+  case KmlNodeType::KmlNetworkLink:
+    type = "KmlNetworkLink";
+    break;
+  case KmlNodeType::KmlPlacemark:
+    type = "KmlPlacemark";
+    break;
+  case KmlNodeType::KmlPhotoOverlay:
+    type = "KmlPhotoOverlay";
+    break;
+  case KmlNodeType::KmlGroundOverlay:
+    type = "KmlGroundOverlay";
+    break;
+  case KmlNodeType::KmlScreenOverlay:
+    type = "KmlScreenOverlay";
+    break;
+  case KmlNodeType::KmlTour:
+    type = "KmlTour";
+    break;
+  default:
+    return "";
+  }
+  return type;
+}
+
+void ListKmlContents::displayChildren(KmlNode* parentNode)
+{
+  if (parentNode == nullptr)
+    return;
+
+  if (KmlContainer* container = dynamic_cast<KmlContainer*>(parentNode))
+  {
+    m_levelNodeNames.clear();
+    bool lastLevel = true;
+
+    // for current level, get names of child nodes
+    for (KmlNode* node: *(container->childNodesListModel()))
+    {
+      QString str = node->name() + " - " + getKmlNodeType(node);
+
+      // if node has children, add ">" to indicate further levels
+      if (!node->children().isEmpty())
+      {
+        str.append(" >");
+        lastLevel = false;
+      }
+      m_levelNodeNames << str;
+    }
+
+    if (lastLevel)
+    {
+      m_currentNode = container->childNodesListModel()->at(0);
+      emit currentNodeChanged();
+    }
+    emit levelNodeNamesChanged();
+  }
+}
+
+// recursively build string to indicate node's ancestors
+void ListKmlContents::buildPathLabel(KmlNode* node)
+{
+  if (node == nullptr)
+    return;
+
+  if (node->parentNode() != nullptr)
+  {
+    buildPathLabel(node->parentNode());
+    m_labelText.append(">");
+  }
+  m_labelText.append(node->name());
+}
+
+void ListKmlContents::displayPreviousLevel()
+{
+  KmlNode* parentNode = m_currentNode->parentNode();
```

Comment:
Intuitively, it makes sense to me that clicking `displayPreviousLevel` should only jump back by one node. Why do we try and skip back by 2?

Reply:
You go back 2 levels to the grandparent and then display its children, so `m_currentNode`'s parent and its siblings